### PR TITLE
1633 add edit and show buttons to ontology versions

### LINKED ISSUE #1633: Add edit button to versions list
For failed ontologies like version 1 in http://ontohub.org/sandbox/syntax%20error/periods///ontology_versions, we want to have an edit button under the red error message box.

This edit button must only be shown if the failed version is the most recent version and if the user is allowed to edit the file. You can see the permission check at the edit button on file view (see the following link).

Clicking this button shall switch the page to the [file view](http://ontohub.org/sandbox/syntax%20error/periods.dol) and act as if the "Edit file" button was clicked on page load.

You can see the corresponding coffeescript of the file view/edit function in [app/assets/javascripts/codemirror-custom.js.coffee](https://github.com/ontohub/ontohub/blob/3aa9e3f744713c1a073dd004fbb77b04ac0b6e0c/app/assets/javascripts/codemirror-custom.js.coffee)
and the edit button in the file view in [app/views/files/show.html.haml:26](https://github.com/ontohub/ontohub/blob/3aa9e3f744713c1a073dd004fbb77b04ac0b6e0c/app/views/files/show.html.haml#L26)

### PR DESCRIPTION
**Pull request by phyrog**
This fixes #1633. Adds edit and show buttons to ontology versions depending on the following factors:

- failed latest version when having write access: edit button
- else: view button to file version at that point

### CODE REVIEW

**Comment by phyrog** [comment]
I will add some cucumber tests for this on monday

File: app/views/ontology_versions/_ontology_version.html.haml [hunk 1]
```diff
@@ -4,4 +4,14 @@
...
+    = link_to 'Download', fancy_repository_path(resource_chain[0], path: ontology_version.path, ref: ontology_version.commit_oid, action: :download), class: 'btn btn-xs btn-default'
+    - latest_version = ontology_version_counter == 0
+    - if can?(:write, ontology.repository) && ontology_version.state == 'failed' && latest_version
+      - btn_text = 'Edit'
+      - anchor = '#edit'
+    - else
+      - btn_text = 'View'
+      - anchor = ''
+    - exact_commit = !latest_version
+    = link_to btn_text, fancy_repository_path(resource_chain[0], path: ontology_version.path, ref: ontology_version.commit_oid, action: :show, exact_commit: exact_commit) + anchor, class: 'btn btn-xs btn-default'
```
Comment (eugenk) [inline app/views/ontology_versions/_ontology_version.html.haml]: Please put the business logic (assigning the variables) of the view into a helper method to keep the view slim and stupid. You could, for instance, put all the code in the `td` except for the download button into a method `edit_or_view_button`.

Comment (phyrog) [inline app/views/ontology_versions/_ontology_version.html.haml]: Alright, done. Had to split it up a lot to get it past hound, though.

File: app/helpers/ontology_version_helper.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
+module OntologyVersionHelper
+  def failed?(ontology_version)
+    ontology_version.state == 'failed'
+  end
+
+  def latest_version?(ontology_version)
+    ontology_version.ontology.versions.last == ontology_version
```
Comment (phyrog) [inline app/helpers/ontology_version_helper.rb]: I'm not sure if it's better to check it this way (was using the iteration variable from the view before). Probably more robust, but definitely slower.

Comment (eugenk) [inline app/helpers/ontology_version_helper.rb]: I agree that this is the most robust version and we should stick with it.

Also, I think we could use this as a method of the model `OntologyVersion`.

File: app/helpers/ontology_version_helper.rb [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+  end
+
+  def can_edit?(ontology, ontology_version)
+    can?(:write, ontology.repository) &&
```
Comment (eugenk) [inline app/helpers/ontology_version_helper.rb]: You don't need the argument `ontology` here. There is an association `repository` on `OntologyVersion`.

File: app/helpers/ontology_version_helper.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+                          path: ontology_version.path,
+                          ref: ontology_version.commit_oid,
+                          action: :show,
+                          exact_commit: !latest_version?(ontology_version))
+  end
+
+  def edit_or_view_button(ontology, ontology_version)
+    btn_text = can_edit?(ontology, ontology_version) ? 'Edit' : 'View'
+    anchor = can_edit?(ontology, ontology_version) ? '#edit' : ''
+
```
Comment (eugenk) [inline app/helpers/ontology_version_helper.rb]: Please cache the result of `can_edit` in a variable instead of calling it twice.

**Comment by phyrog** [comment]
Ok, I wrote some tests for the buttons, but somehow after opening the edit button, the File browser is shown (instead of the view/edit page). `current_path` shows the correct path, but `save_and_open_page` definitely shows the wrong page.

**Comment by eugenk** [comment]
I think that's because there is no file for the ontology in the git repository. Our factories skip the file creation because we don't need them most of the time. You can add a file for the ontology with something like
```ruby
  repository = @ontology.repository
  user = repository.permissions.first.subject
  Tempfile.create('testfile') do |f|
    repository.save_file_only(f.path, @ontology.path, 'add ontology file', user)
  end
```

File: features/OntologyVersions.feature [hunk 5]
```diff
@@ -18,6 +18,7 @@ Scenario: Displaying an edit button on the latest ontology if it is failed with
   And I am logged in
   And there is an ontology with a "failed" version
   And I have permissions to edit the ontology
+  And there is a ontology file
```
Comment (eugenk) [inline features/OntologyVersions.feature]: typo: there is *an* ontology file ;)

**Comment by eugenk** [comment]
:+1:

**Pull request merged**